### PR TITLE
[test] modernize ecdsatest and extend ECDSA sign KATs

### PR DESCRIPTION
**Pull request by bbbrumley**
Follow up to my #8232 promises.

This extends ECDSA sign KATs from two (rare) curves and having two X9.62 KATs each to 724 KATs from [NIST CAVP](https://csrc.nist.gov/Projects/Cryptographic-Algorithm-Validation-Program/Digital-Signatures) spanning all NIST prime and binary curves with at least 224 bits.

I got a little heavy handed since it looks like `ecdsatest` hasn't really been significantly touched since the coding guidelines and the new test framework went into place.

For the non-KAT tests, switches from old `ECDSA_sign` interface to `EVP`. (I usually don't operate on this level of the lib so any comments there are welcome.)

Also picks up some extra coverage with key generation KATs along the way -- they were already a part of NIST CAVP ECDSA KATs so I threw it in.

Maybe master and 1.1.1? Tagging @romen for review.

I will eventually have another follow up PR for `evp_test` to pick up the verification KATs from NIST CAVP.

### CODE REVIEW

File: test/ecdsatest.c [hunk 1]
```diff
@@ -8,32 +8,24 @@
...
-#  include <openssl/engine.h>
-# endif
-# include <openssl/sha.h>
-# include <openssl/err.h>
-# include <openssl/rand.h>
+#include <openssl/evp.h>
+#include <openssl/bn.h>
+#include <openssl/ec.h>
+#include <openssl/rand.h>
+#include "ecdsatest.h"
```
Comment (romen) [inline test/ecdsatest.c]: Nit: These are missing a space, i.e. `# include`

File: test/ecdsatest.c [hunk 2]
```diff
@@ -57,343 +49,269 @@ static int restore_rand(void)
...
-        wrong_eckey = NULL;
-        ECDSA_SIG_free(ecdsa_sig);
-        ecdsa_sig = NULL;
-        ECDSA_SIG_free(modified_sig);
-        modified_sig = NULL;
-        OPENSSL_free(raw_buf);
-        raw_buf = NULL;
+    /* skip built-in curves where ord(G) is not prime */
+    if (nid == NID_ipsec4 || nid == NID_ipsec3) {
+        TEST_info("ECDSA unsupported for curve %s", OBJ_nid2sn(nid));
```
Comment (romen) [inline test/ecdsatest.c]: optional nit: maybe add an explicit `skipped: ` at the beginning of the message?

**Comment by kroeckx** [comment]
The compilers generates:
```
test/ecdsatest.c: In function 'x9_62_tests':
test/ecdsatest.c:174:16: error: 'qbuf' may be used uninitialized in this function [-Werror=maybe-uninitialized]
     OPENSSL_free(qbuf);
                ^
test/ecdsatest.c:101:27: note: 'qbuf' was declared here
     unsigned char *pbuf, *qbuf, *message = NULL;
                           ^
test/ecdsatest.c: In function 'test_builtin':
test/ecdsatest.c:311:19: error: 'pkey_neg' may be used uninitialized in this function [-Werror=maybe-uninitialized]
      EVP_PKEY_free(pkey_neg);
                   ^
test/ecdsatest.c:206:15: note: 'pkey_neg' was declared here
     EVP_PKEY *pkey_neg, *pkey = NULL;
               ^
cc1: all warnings being treated as errors
```

File: test/ecdsatest.c [hunk 3]
```diff
@@ -402,8 +320,20 @@ int setup_tests(void)
...
 #else
-    ADD_TEST(x9_62_tests);
-    ADD_TEST(test_builtin);
+    /* get a list of all internal curves */
+    crv_len = EC_get_builtin_curves(NULL, 0);
+    if (!TEST_ptr(curves = OPENSSL_malloc(sizeof(*curves) * crv_len))
+        || !TEST_true(EC_get_builtin_curves(curves, crv_len)))
+        return 0;
+    ADD_ALL_TESTS(test_builtin, crv_len);
+    ADD_ALL_TESTS(x9_62_tests, sizeof(ecdsa_cavs_kats)/sizeof(ecdsa_cavs_kat_t));
```
Comment (romen) [inline test/ecdsatest.c]: Probably better to use `OSSL_NELEM()` here?

File: test/ecdsatest.h [hunk 4]
```diff
@@ -0,0 +1,10216 @@
...
+ */
+
+#ifndef ECDSATEST_CAVS_H
+#define ECDSATEST_CAVS_H
+
+/*-
+ * NIST CAVP ECDSA KATs:
+ * https://csrc.nist.gov/CSRC/media/Projects
+ * /Cryptographic-Algorithm-Validation-Program
+ * /documents/dss/186-3ecdsatestvectors.zip
```
Comment (romen) [inline test/ecdsatest.h]: nit: I believe the URL should stay in one line, even if it defies the style guidelines

File: test/ecdsatest.c [hunk 5]
```diff
@@ -57,343 +49,269 @@ static int restore_rand(void)
...
+ * - accept those signatures as valid
+ */
+static int x9_62_tests(int n)
 {
-    int ret = 0;
-    const char message[] = "abc";
-    unsigned char digest[SHA_DIGEST_LENGTH];
+    int nid, md_nid, ret = 0;
+    const char *r_in, *s_in, *tbs = NULL;
+    unsigned char *pbuf, *qbuf, *message = NULL;
```
Comment (romen) [inline test/ecdsatest.c]: As catched by the CI and mentioned by @kroeckx , `pbuf` and `qbuf` should also initialized to `NULL`

Comment (bbbrumley) [reply test/ecdsatest.c]: What's the magic `config` option to make my own build barf on this? To test before I fixup.

File: test/ecdsatest.c [hunk 6]
```diff
@@ -57,343 +49,269 @@ static int restore_rand(void)
...
-    unsigned char *sig_ptr2;
-    unsigned char *raw_buf = NULL;
-    const BIGNUM *sig_r, *sig_s;
-    BIGNUM *modified_r = NULL, *modified_s = NULL;
-    BIGNUM *unmodified_r = NULL, *unmodified_s = NULL;
-    unsigned int sig_len, order, r_len, s_len, bn_len, buf_len;
+    EC_KEY *eckey_neg, *eckey = NULL;
+    unsigned char tbs[128];
+    unsigned char *sig = NULL;
+    EVP_PKEY *pkey_neg, *pkey = NULL;
```
Comment (romen) [inline test/ecdsatest.c]: `pkey_neg` should also be initialized as `NULL`

File: test/ecdsatest.c (hunk 5 above)
Comment (romen) [reply test/ecdsatest.c]: It's compiling through the clang in Xcode 9.3

Comment (romen) [reply test/ecdsatest.c]: and with `--strict-warnings`

**Comment by bbbrumley** [comment]
> The compilers generates:

Thanks! fixup in adb59a11447dbcb320a35ab395e0c863756afa5c

File: test/ecdsatest.c (hunk 5 above)
Comment (bbbrumley) [reply test/ecdsatest.c]: Thanks! fixup in adb59a11447dbcb320a35ab395e0c863756afa5c

File: test/ecdsatest.c [hunk 7]
```diff
@@ -8,32 +8,25 @@
...
  */
 
-#include <stdio.h>
-#include <stdlib.h>
-#include <string.h>
-
-#include <openssl/opensslconf.h> /* To see if OPENSSL_NO_EC is defined */
-#include "testutil.h"
+# include <openssl/opensslconf.h> /* To see if OPENSSL_NO_EC is defined */
+# include "testutil.h"
```
Comment (romen) [inline test/ecdsatest.c]: nit: these before the `#ifndef` should not be indented with the extra space, i.e. `#include`

File: test/ecdsatest.c [hunk 8]
```diff
@@ -57,343 +50,269 @@ static int restore_rand(void)
...
+    static int fbytes_counter = 0;
     BIGNUM *tmp = NULL;
 
     if (use_fake == 0)
         return old_rand->bytes(buf, num);
 
     use_fake = 0;
 
-    if (fbytes_counter >= 8)
+    if (fbytes_counter >= 2)
```
Comment (FdaSilvaYY) [inline test/ecdsatest.c]: Probably better to use `OSSL_NELEM()` instead of `2` ?

Comment (bbbrumley) [reply test/ecdsatest.c]: Thanks! fixed in c51ffb56485860f631540b5b329397eefff23bf6

File: test/ecdsatest.c (hunk 7 above)
Comment (bbbrumley) [reply test/ecdsatest.c]: Thanks! cherry picked some stuff from `openssl-format-source`, then fixed in c51ffb56485860f631540b5b329397eefff23bf6

**Comment by bbbrumley** [comment]
Couple of questions/concerns from my side.

1. I don't understand how `change_rand` and `restore_rand` are working at all. ECC keygen and ECDSA sign use _lots_ of randomness (e.g. projective blinding), not just for the private key and nonce. How is the app/lib magically sorting these out?
2. This is my first time using the newish `EVP_DigestSign` and `EVP_DigestVerify` -- am I doing it right? I couldn't get negative tests to fail unless I called `EVP_MD_CTX_reset` between. That behavior seemed a little odd from the application developer perspective; e.g. below would fail the negative test on the last line, returning `1`:

```
/* positive test, verify with correct key, 1 return */
|| !TEST_true(EVP_DigestVerifyInit(mctx, NULL, NULL, NULL, pkey))
|| !TEST_int_eq(EVP_DigestVerify(mctx, sig, sig_len, tbs, sizeof(tbs)), 1))
/* negative test, verify with wrong key, 0 return */
|| !TEST_true(EVP_DigestVerifyInit(mctx, NULL, NULL, NULL, pkey_neg))
|| !TEST_int_eq(EVP_DigestVerify(mctx, sig, sig_len, tbs, sizeof(tbs)), 0)
```

**Comment by romen** [comment]
> 1. I don't understand how `change_rand` and `restore_rand` are working at all. ECC keygen and ECDSA sign use _lots_ of randomness (e.g. projective blinding), not just for the private key and nonce. How is the app/lib magically sorting these out?

`change_rand` is saving the default `RAND_METHOD` in `old_rand`, and replacing it with `fake_rand`, a clone of it that replaces only `.bytes` with the local `fbytes` function.

`fbytes` only returns predetermined data the first time it's called since setting `use_fake != 0`: if `use_fake == 0` it invokes `old_rand->bytes`, otherwise it sets `use_fake=0` before returning predetermined data. 

`restore_rand` registers again the saved `old_rand` as the library `RAND_METHOD`, but functionally, apart from an extra indirection, using `fake_rand` when `use_fake == 0` is equivalent to using `old_rand`.

Admittedly, `x9_62_tests` seems fragile, as we are implicitly assuming that we need to serve predetermined data exclusively and exactly to the first request of randomness after we manually set `use_fake = 1`, i.e. we assume that `EC_KEY_generate_key` and `ECDSA_sign_setup` won't require other *randomness* before the bytes we want to provide deterministically.

> 2. This is my first time using the newish `EVP_DigestSign` and `EVP_DigestVerify` -- am I doing it right? I couldn't get negative tests to fail unless I called `EVP_MD_CTX_reset` between. That behavior seemed a little odd from the application developer perspective; e.g. below would fail the negative test on the last line, returning `1`:
> 
> ```
> /* positive test, verify with correct key, 1 return */
> || !TEST_true(EVP_DigestVerifyInit(mctx, NULL, NULL, NULL, pkey))
> || !TEST_int_eq(EVP_DigestVerify(mctx, sig, sig_len, tbs, sizeof(tbs)), 1))
> /* negative test, verify with wrong key, 0 return */
> || !TEST_true(EVP_DigestVerifyInit(mctx, NULL, NULL, NULL, pkey_neg))
> || !TEST_int_eq(EVP_DigestVerify(mctx, sig, sig_len, tbs, sizeof(tbs)), 0)
> ```

It's documented in [EVP_DigestVerifyInit](https://github.com/openssl/openssl/blob/master/doc/man3/EVP_DigestVerifyInit.pod) and [EVP_DigestSignInit](https://github.com/openssl/openssl/blob/master/doc/man3/EVP_DigestSignInit.pod):

> `ctx` must be created with `EVP_MD_CTX_new()` before calling this function.

The reason being related to the fact that when passing a `ctx` that has not been already assigned an externally created `EVP_PKEY_CTX`, an
> `EVP_PKEY_CTX` is created inside `EVP_DigestVerifyInit()` and it will be freed automatically when the `EVP_MD_CTX` is freed.

>No `EVP_PKEY_CTX` will be created by `EVP_DigestVerifyInit()` if the passed `ctx `has already been assigned one via `EVP_MD_CTX_set_ctx(3)`. See also `SM2(7)`.

Calling `EVP_MD_CTX_reset` is technically equivalent to calling `EVP_MD_CTX_free()` minus actually freeing the memory for the `EVP_MD_CTX` struct, all the other internal resources are cleared and freed, including the tracked `EVP_PKEY_CTX` (if it wasn't provided externally by the caller of `EVP_DigestVerifyInit()`. 

P.S. this was a good exercise, I spotted a few typos in the `pod` files, PR coming soon.

**Comment by richsalz** [comment]
Thanks, @bbbrumley this is nice work!

**Comment by bbbrumley** [comment]
> `ctx` must be created with `EVP_MD_CTX_new()` before calling this function.

Yea, I read that when the tests were failing. I just don't think it's very useful documentation. Because my brain said "of course it came from `EVP_MD_CTX_new` -- where else would it come from?" ( It is already there on L219 in the PR.) I think it is trying to imply `ctx` needs to be in some kind of pristine-like state; e.g. it is OK to call A B C but not X Y Z. I just don't know what A B C and X Y Z are, and stumbled through it in this PR to get the tests passing.

File: test/ecdsatest.h [hunk 9]
```diff
@@ -0,0 +1,10214 @@
...
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef ECDSATEST_CAVS_H
+#define ECDSATEST_CAVS_H
```
Comment (mattcaswell) [inline test/ecdsatest.h]: Style nit: Should be "# define" (i.e with a space)

File: test/ecdsatest.h [hunk 10]
```diff
@@ -0,0 +1,10214 @@
...
+    "9886ae46c1415c3bc959e82b760ad760aab66885a84e620aa339fdf102465c42",
+    "2bf3a80bc04faa35ebecc0f4864ac02d349f6f126e0f988501b8d3075409a26c"},
+    {NID_X9_62_prime256v1, NID_sha256,
+    "3b6ee2425940b3d240d35b97b6dcd61ed3423d8e71a0ada35d47b322d17b35ea0472f35e"
+    "dd1d252f87b8b65ef4b716669fc9ac28b00d34a9d66ad118c9d94e7f46d0b4f6c2b2d339"
+    "fd6bcd351241a387cc82609057048c12c4ec3d85c661975c45b300cb96930d89370a327c"
+    "98b67defaa89497aa8ef994c77f1130f752f94a4",
+    "d4234bebfbc821050341a37e1240efe5e33763cbbb2ef76a1c79e24724e5a5e7",
+    "048fb287f0202ad57ae841aea35f29b2e1d53e196d0ddd9aec24813d64c0922fb71f6daf"
+    "f1aa2dd2d6d3741623eecb5e7b612997a1039aab2e
```
Comment (mattcaswell) [inline test/ecdsatest.h]: "# ifndef"

Comment (mattcaswell) [inline test/ecdsatest.h]: "# endif"

Comment (bbbrumley) [reply test/ecdsatest.h]: Thanks! Missed `openssl-format-source` on the header file -- fixup in 60bcac5e3ffa8f625373b64b7523d2dcc022d93d

**Comment by romen** [comment]
Merged to `master` (1a31d8017ee7e8df0eca76fee601b826699c9ac1) and `1.1.1` (555b2593f7e0d3376043aa5f70d657d29e2f8ad8).

Thanks @bbbrumley for the contribution and all for the feedback!

**Pull request closed without merging**